Estimate Next Purchase

sanrafa opened the pull request:
## Description
In this pull request, we estimate the days until the next purchase.

_Data flow_

1. When the user marks an item as purchased, 
2. updateDocument() is fired (which in addition to setting the purchased date, now called last_purchased_date),
3. increments the total_purchases field, 
4. stores a days_since_last_transaction, and estimated_next_purchase field by calling updateEstimate(),
5. triggering the calculateEstimate() imported with the new data points.

_Firestore database updated schema_

![Screen Shot 2022-02-10 at 3 30 38 PM](https://user-images.githubusercontent.com/67841237/153491572-4c32d462-c171-488a-b3d7-dede5d107a4e.png)



<!-- What does this code change? Why did I choose this approach? Did I learn anything worth sharing? Reminder: This will be a publicly facing representation of your work (READ: help you land that sweet dev gig). -->

## Related Issue

<!-- If you write "closes" followed by the Github issue number, it will automatically close the issue for you when the PR merges -->

Closes #10 

## Acceptance Criteria

- [X] When a purchase is recorded, the estimated number of days until the next purchase date should be calculated and recorded in the database
- [X] The script calculateEstimate from @the-collab-lab/shopping-list-utils should be used to calculate the next estimated purchase interval. 

 <!-- Include AC from the Github issue -->

## Type of Changes

<!-- Put an `✓` for the applicable box: -->

|     | Type                       |
| --- | -------------------------- |
| ✓   | :sparkles: New feature     |
|✓ | :link: Update dependencies |


## calculateEstimate

[calculateEstimate](https://github.com/the-collab-lab/shopping-list-utils/blob/main/src/calculateEstimate/index.ts)

_Deep dive into how the imported function estimates the next purchase date._


<!-- If UI feature, take provide screenshots -->

### Video Walkthrough

https://user-images.githubusercontent.com/67841237/153494694-01ddade3-fa0e-45d9-8fc0-54753634c59f.mp4


<!-- If UI feature, take provide screenshots -->

## Testing Steps / QA Criteria

<!-- Provide steps the other cohort members and mentors need to follow to properly test your additions. -->
1. `git checkout dp-rb-estimate-purchase` Checkout the branch.
2. `npm install` (we added the calculateEstimate package)
3.  `npm start` (fire up the server)
4. Join the list with token `nee calm social`
5.  Mark an item purchased, you should expect to see the total purchases increment, last purchase date update, and next purchase date.
6. In order to test if anything is being recalculated, you would need to go into the Firestore database for that list token, find the item, and manipulate the last purchased date to the past, so you can purchase it again. 
[Firestore db](https://console.firebase.google.com/project/tcl-37-smart-shopping-list/firestore/data/~2Faid%20misty%20onus~2FJLJ6WxYxP9bwSleQYcC3)
(See video walkthrough)

sanrafa committed:
import calculateEstimate from collab-lab utils

danainjax committed:
added fields to firestore

sanrafa committed:
Calc days since last transaction

sanrafa committed:
Edit db schema & add updateEstimate helper func

danainjax committed:
added attributes to the ui so we could see what values the db was returning

danainjax committed:
conditionally rendering the last purchased date and estimated next purchase, changed field in db to estimated_next_purchase for clarity, partners agreed

danainjax committed:
added function to parse the date in the last purchased date ui

danainjax committed:
added spaces

yenly commented on src/components/List.jsx:
```diff
@@ -43,14 +93,29 @@ export const List = ({ token }) => {
                 onChange={(e) => handleClick(doc, e)}
               />
               <label htmlFor={doc.id}>{doc.data().item}</label>
+              {doc.data().total_purchases > 0 ? (
+                <p> Total purchases: {doc.data().total_purchases}</p>
+              ) : null}
```
Not a must change. For future ref, this is a common pattern to use when your conditional for false is null or do nothing.
```suggestion
              {doc.data().total_purchases > 0 && <p> Total purchases: {doc.data().total_purchases}</p>}
```
More info on [MDN Logical AND](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Logical_AND)

danainjax replied:
Thanks Yenly. Good call.

danainjax replied:
I see, the short circuit &&. That took a moment. Thank you!

danainjax replied:
Committed the changes

yenly reviewed (approved):
Well done on this complicated issue! The video walk-through was very clear and helpful. I would like for us as a team to discuss the changes in the db. Was it necessary to introduce all the extra fields? Can this feature still work with less?

vividavila98 commented on src/components/List.jsx:
```diff
@@ -20,17 +21,66 @@ export const List = ({ token }) => {
 
   const handleClick = (data, e) => {
     updateDocument(data.id);
   };
```
I'm wondering if we pass in the entire docment object (called `data` here) instead of just its id to `updateDocument()`, could we set `docData` equal to `document.data()` without needing to use `getDoc()`? This would only remove a couple lines, so it's not a necessary change!

danainjax replied:
Good suggestion. Looking at this one to implement.

vividavila98 reviewed (approved):
I learned so much looking over this! Great job as always :) Lmk if my comment makes sense!

tannaurus commented on src/components/List.jsx:
```diff
@@ -20,17 +21,66 @@ export const List = ({ token }) => {
+        total_purchases,
+      } = docData;
+      // if no estimated_next_purchase exists, use 'undefined' in calculateEstimate (cannot save undefined fields in db)
+      estimated_next_purchase
```
if we move the declaration of `prevEstimate` down a bit we can improve the readability of this function
```suggestion
let prevEstimate = undefined;
if (estimated_next_purchase) {
      prevEstimate = estimated_next_purchase
 }
```

tannaurus commented on src/components/List.jsx:
```diff
@@ -20,17 +21,66 @@ export const List = ({ token }) => {
+      docData = document.data();
+      console.log(docData);
+    }
+    const updateEstimate = () => {
```
A quick comment on function names: generally when I read "updateX" I assume it might be making an api request or mutating something, `getX` tells me it is going to go off and get a new value which I feel is more applicable here 

```suggestion
    const getEstimate = () => {
```

tannaurus commented on src/components/List.jsx:
```diff
@@ -20,17 +21,66 @@ export const List = ({ token }) => {
 
+  const calcDaysSince = (transactionDate) => {
+    const timeNow = moment().format();
+    const date1 = moment(timeNow, 'YYYYMMDD HH:mm:ss');
```
This could be just

```suggestion
    const date1 = moment();
```

danainjax replied:
Light dawns!

tannaurus commented on src/components/List.jsx:
```diff
@@ -20,17 +21,66 @@ export const List = ({ token }) => {
+  };
+
+  const cleanDate = (date) => {
+    const lastPurchasedDate = moment(date).format('MMMM Do, YYYY');
```
something to consider when writing a function like this: we started off on a great foot by keeping the function name generic: "cleanDate", but the variables within the function indicate that this should be used for a specific purpose, "lastPurchasedDate" - this may confuse another dev should they come in here and want to work with this function, especially if it were to be longer than it is right now. A little change to something like "formattedDate" would keep this function generic 

```suggestion
    const formattedDate = moment(date).format('MMMM Do, YYYY');
```

danainjax committed:
Update src/components/List.jsx

Co-authored-by: Tanner Gill <contacttannergill@gmail.com>

danainjax committed:
updated to getEstimate()

danainjax committed:
Update src/components/List.jsx

Co-authored-by: Tanner Gill <contacttannergill@gmail.com>

danainjax committed:
renamed formattedDate in cleanDate function for more useability

danainjax committed:
Update src/components/List.jsx

Co-authored-by: Tanner Gill <contacttannergill@gmail.com>

danainjax committed:
removed timeNow variable, already capturing with moment()

danainjax committed:
removed null from the condtional rendering of estimated next purchase and used && short circuit instead

sanrafa committed:
Change arg for updateDocument to document instead of document.id

sanrafa committed:
Move calc functions to helpers.js & import

sanrafa commented:
Thought the component was getting quite cluttered, so I moved the non-DB calc/logic functions to `helpers.js` and re-imported them. Also removed the "days_since_last_transaction" field from the DB schema as it could be calculated in getEstimate(). 

Thank you @yenly @tannaurus @vividavila98 for the feedback!
